modules/placement-driver/src/main/java/org/apache/ignite/internal/raft/client/TopologyAwareRaftGroupService.java

```diff
@@ -0,0 +1,443 @@
+     * @param configuration Group configuration.
+     * @param getLeader True to get the group's leader upon service creation.
+     * @param executor RPC executor.
+     * @param logicalTopologyService Logical topology service.
+     * @return Future to create a raft client.
+     */
+    public static CompletableFuture<RaftGroupService> start(
+            ReplicationGroupId groupId,
+            ClusterService cluster,
+            RaftMessagesFactory factory,
+            RaftConfiguration raftConfiguration,
+            PeersAndLearners configuration,
+            boolean getLeader,
+            ScheduledExecutorService executor,
+            LogicalTopologyService logicalTopologyService
+    ) {
+        return RaftGroupServiceImpl.start(groupId, cluster, factory, raftConfiguration, configuration, getLeader, executor)
+                .thenApply(raftGroupService -> new TopologyAwareRaftGroupService(cluster, factory, executor, raftConfiguration,
+                        raftGroupService, logicalTopologyService));
+    }
+
+    /**
+     * Sends a subscribe message to a specific node of the cluster.
+     *
+     * @param node Node.
+     * @param msg  Subscribe message.
+     * @return A future that will complete when the message is sent.
+     */
+    private CompletableFuture<Boolean> sendSubscribeMessage(ClusterNode node, SubscriptionLeaderChangeRequest msg) {
+        var msgSendFut = new CompletableFuture<Boolean>();
+
+        sendWithRetry(node, msg, msgSendFut);
+
+        return msgSendFut;
+    }
+
+    /**
+     * Tries to send a subscribe message until the node leaves of the cluster.
+     *
+     * @param node       Node.
+     * @param msg        Subscribe message to send.
+     * @param msgSendFut Future that will completed when the message is send or an issue prevent that.
+     */
+    private void sendWithRetry(ClusterNode node, SubscriptionLeaderChangeRequest msg, CompletableFuture<Boolean> msgSendFut) {
+        clusterService.messagingService().invoke(node, msg, raftConfiguration.responseTimeout().value()).whenCompleteAsync((unused, th) -> {
+            if (th != null) {
+                if (recoverable(th)) {
+                    logicalTopologyService.logicalTopologyOnLeader().whenCompleteAsync((logicalTopologySnapshot, topologyGetIssue) -> {
+                        if (topologyGetIssue != null) {
+                            LOG.error("Actual logical topology snapshot was not get.", topologyGetIssue);
+
+                            msgSendFut.completeExceptionally(topologyGetIssue);
+
+                            return;
+                        }
+
+                        if (logicalTopologySnapshot.nodes().contains(node)) {
+                            sendWithRetry(node, msg, msgSendFut);
+                        } else {
+                            LOG.info("Could not subscribe to leader update from a specific node, because the node had left of the cluster"
```

Comment:
```suggestion
                            LOG.info("Could not subscribe to leader update from a specific node, because the node had left from the cluster"
```